### PR TITLE
PUBDEV-8742: Using the model performance API to avoid unnecessary transforms in the scikit-learn wrapper

### PR DESCRIPTION
**Pull request by sebhrusen**
https://h2oai.atlassian.net/browse/PUBDEV-8742

migration of https://github.com/h2oai/h2o-3/pull/6231 into `h2o-3` repo so it can go through our Jenkins pipeline

```
* Adding support for the model performance API in the scikit-learn wrappers.

Co-authored-by: Nikhil Arora <nikhiljarora@outlook.com>

* Fixing issues brought up in PR review

Co-authored-by: Nikhil Arora <nikhiljarora@outlook.com>
```

### CODE REVIEW

**Review by michalkurka — APPROVED** [review]

**Comment by sebhrusen** [comment]
@ak-gupta @NikhilJArora , I can't merge those changes as they create several issues due to fundamental differences between `sklearn` and `H2O` metrics:
- in `H2O`, `model_performance().accuracy()` is available only for binary tasks, not for multiclass.
- `model_performance().accuracy()` returns a list of pairs (by default one pair only for the threshold maximising F1 metric): [[threshold, accuracy(threshold)]] 
- even if using `model_performance().accuracy()[0][1]` to extract the accuracy value, this value can be different from the `sklearn` accuracy due to the H2O contract: see https://docs.h2o.ai/h2o/latest-stable/h2o-docs/performance-and-prediction.html#accuracy
```
For binary classification problems, H2O uses the model along with the given dataset to calculate the threshold that will give the maximum F1 for the given dataset.
```

as those wrappers are designed to work in a `sklearn` context, I think we should keep using the `sklearn` metrics everywhere for consistency and clarity.

**Comment by ak-gupta** [comment]
Ah ok, thanks for the deep dive @sebhrusen. For other users that look at this PR and want to use the `scikit-learn` wrapper with a meta-estimator, we have been using a custom scorer:

```python
def h2o_auc_scorer(estimator, X, y, **metric_kwargs) -> float:
    """Sklearn scorer function for h2o model ``auc``.

    Wrapper function for
    ``h2o.model.metrics_base.H2OBinomialModelMetrics.auc``.

    Parameters
    ----------
    estimator : h2o.estimators
        Estimator object that is not using h2o sklearn-wrapper.
    X: ndarray or DataFrame
        Data on which score will be computed.
    y: array-like
        Target column.
    **metric_kwargs : key, value
        Key word arguments to pass to
        ``h2o.model.metrics_base.H2OBinomialModelMetrics.auc``.

    Returns
    -------
    float
        `auc` score result.
    """
    if not hasattr(estimator, "model_performance") and hasattr(estimator, "_estimator"):
        estimator = estimator._estimator
    scoring_func = estimator.model_performance(X.cbind(y))
    score: float = scoring_func.auc(**metric_kwargs)

    return score
```

and supplying it to the `scoring` initialization parameter.

**Comment by sebhrusen** [comment]
@ak-gupta thanks for the custom scorer example, I think it's definitely a better approach if you really need to use `model_performance`, also thanks to `sklearn` for allowing those kind of customizations.

Closing this PR then.
